File: yaf-gadomain/src/main/java/ar/edu/ungs/yamiko/ga/domain/impl/BasicPhenotype.java
package ar.edu.ungs.yamiko.ga.domain.impl;

import java.io.Serializable;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import ar.edu.ungs.yamiko.ga.domain.Chromosome;
import ar.edu.ungs.yamiko.ga.domain.Gene;
import ar.edu.ungs.yamiko.ga.domain.Phenotype;

/**
 * Implementación Básica de Fenotipo
 * @author ricardo
 *
 */
public class BasicPhenotype implements Phenotype,Serializable{

	/**
	 * 
	 */
	private static final long serialVersionUID = 4838290437952651077L;
	@SuppressWarnings("rawtypes")
	private Map<Chromosome, Map<Gene, Object>> alleleMap;
	
	@SuppressWarnings("rawtypes")
	public BasicPhenotype(Chromosome c, Map<Gene, Object> _alleles) {
		alleleMap=new HashMap<Chromosome, Map<Gene,Object>>();
		alleleMap.put(c, _alleles);
	}
	@SuppressWarnings("rawtypes")
	public Map<Chromosome, Map<Gene, Object>> getAlleleMap() {
		return alleleMap;
	}
	
	public Collection<Map<Gene, Object>> getAlleles() {
		return alleleMap.values();
	}
	

}

File: yaf-gadomain-scala/src/main/scala/ar/edu/ungs/yamiko/ga/domain/impl/BasicPhenotype.scala
package ar.edu.ungs.yamiko.ga.domain.impl

import ar.edu.ungs.yamiko.ga.domain.Gene
import ar.edu.ungs.yamiko.ga.domain.Genome
import ar.edu.ungs.yamiko.ga.domain.Ribosome
import ar.edu.ungs.yamiko.ga.domain.Genotype
import ar.edu.ungs.yamiko.ga.domain.Individual
import ar.edu.ungs.yamiko.ga.domain.Phenotype
import ar.edu.ungs.yamiko.ga.domain.Chromosome
import ar.edu.ungs.yamiko.ga.domain.Phenotype

@SerialVersionUID(1199L)
class BasicPhenotype[T](chromosome:Chromosome[T], alleles:Map[Gene, T] ) extends Phenotype[T]{
  
  private val alleleMap:Map[Chromosome[T], Map[Gene, T]]=Map(chromosome -> alleles)
  
	override def getAlleleMap()=alleleMap
	override def getAlleles():List[Map[Gene, T]]=alleleMap.values.toList

  def canEqual(a:BasicPhenotype[T]) = a.isInstanceOf[BasicPhenotype[T]]

  override def equals(that: Any): Boolean =
    that match {
      case that: BasicPhenotype[T] => 
        that.canEqual(this) && this.hashCode == that.hashCode
      case _ => false
  }

  override def hashCode:Int = {
    val ourHash = if (chromosome == null) 0 else chromosome.hashCode 
    if (alleles == null) super.hashCode + ourHash else super.hashCode + ourHash   + alleles.hashCode()     
  }
  
}